### PR TITLE
Fix avg_over_time for nan and float64 overflows

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

**Comment by roidelapluie** [comment]
This does not protect from float64 overflows like the old function was so that is not a real fix

**Comment by brian-brazil** [comment]
Can you explain more what this is trying to do, and if/how it varies from the naieve algorithm?

**Comment by roidelapluie** [comment]
It fixes avg_over_time for Inf + avg_over_time for float64 that would otherwise overflow the float64.

e.g. the following tests would fail otherwise:

```
  metric5 1 2 3 Inf Inf
  metric10 -9.988465674311579e+307 9.988465674311579e+307

eval instant at 1m avg_over_time(metric5[1m])
  {} Inf

eval instant at 1m avg_over_time(metric10[1m])
  {} 0
```

**Comment by roidelapluie** [comment]
I will make the comments cleaerer

**Comment by brian-brazil** [comment]
What are those returning now? Does `sum_over_time / count_over_time == avg_over_time`? Would the avg aggregator need changes too?

**Comment by roidelapluie** [comment]
## Current behaviour

```
load 10s
  metric5 1 2 3 Inf Inf
  metric6 1 2 3 -Inf -Inf
  metric7 1 2 -Inf -Inf Inf
  metric10 -9.988465674311579e+307 9.988465674311579e+307

eval instant at 1m avg_over_time(metric5[1m])
  {} Nan

eval instant at 1m avg_over_time(metric6[1m])
  {} Nan

eval instant at 1m avg_over_time(metric10[1m])
  {} +Inf
```

## New

```
load 10s
  metric5 1 2 3 Inf Inf
  metric6 1 2 3 -Inf -Inf
  metric7 1 2 -Inf -Inf Inf
  metric10 -9.988465674311579e+307 9.988465674311579e+307

eval instant at 1m avg_over_time(metric5[1m])
  {} Inf

eval instant at 1m avg_over_time(metric6[1m])
  {} -Inf

eval instant at 1m avg_over_time(metric10[1m])
  {} 0
```

**Comment by brian-brazil** [comment]
Why is metric10 changing?

**Comment by roidelapluie** [comment]
It is one big positive and one big negative float that average to 0

**Comment by roidelapluie** [comment]
Previously (v.V - mean ) / count would overlow to (Inf) / count => Int

**Comment by roidelapluie** [comment]
avg_over_time is the same except for metric9 and metric8 which overflows float64.

**Comment by roidelapluie** [comment]
`avg_over_time`  is the same as `sum_over_time / count_over_time` except for metric9 and metric8 which overflows float64.

**Comment by roidelapluie** [comment]
@brian-brazil are we interested by this fix?

File: promql/engine.go [hunk 1]
```diff
@@ -2002,7 +2002,22 @@ func (ev *evaluator) aggregation(op parser.ItemType, grouping []string, without
...
+			if math.IsInf(group.mean, 0) {
+				if math.IsInf(s.V, 0) && (group.mean > 0) == (s.V > 0) {
+					// The `mean` and `s.V` values are `Inf` of the same sign.  They
+					// can't be substracted, but the value of `mean` is correct
+					// already.  Continue the loop. `break` can not be used because
+					// if an opposite `Inf` is met later, `mean` must become `NaN`.
+					break
+				}
+				if !math.IsInf(s.V, 0) && !math.IsNaN(s.V) {
+					// It the mean is an `Inf`, avoid removing an `Inf` with
```
Comment (brian-brazil) [inline promql/engine.go:2014]: I don't understand this comment.

**Comment by roidelapluie** [comment]
I have adapted the comment and rebased on top of group().

File: promql/engine.go [hunk 2]
```diff
@@ -2005,7 +2005,25 @@ func (ev *evaluator) aggregation(op parser.ItemType, grouping []string, without
...
-			group.mean += (s.V - group.mean) / float64(group.groupCount)
+			if math.IsInf(group.mean, 0) {
+				if math.IsInf(s.V, 0) && (group.mean > 0) == (s.V > 0) {
+					// The `mean` and `s.V` values are `Inf` of the same sign.  They
+					// can't be subtracted, but the value of `mean` is correct
+					// already.
+					break
+				}
+				if !math.IsInf(s.V, 0) && !math.IsNaN(s.V) {
+					// At this stage, the mean is an infinite. It the added
```
Comment (brian-brazil) [inline promql/engine.go:2016]: If the

**Pull request merged**